dcap/types/src/tcb.rs

```diff
@@ -0,0 +1,362 @@
+    Der(x509_cert::der::Error),
+    FmspcSize(usize),
+}
+
+impl From<x509_cert::der::Error> for Error {
+    fn from(err: x509_cert::der::Error) -> Self {
+        Error::Der(err)
+    }
+}
+
+// The SGX extensions aren't really documented. They aren't RFC 5280 extensions
+// which are `OID` and `OCTET`. They're an `OID` and an `Any`, which is what the
+// [`AttributeTypeAndValue`] is.
+type SgxExtensions = Vec<AttributeTypeAndValue>;
+
+/// The TCB info provided by the PCK(Provisioning Certification Key) leaf
+/// certificate
+#[derive(Debug, PartialEq)]
+pub struct PckTcbInfo {
+    svns: [u32; 16],
+    pce_svn: u32,
+    fmspc: [u8; FMSPC_SIZE],
+}
+
+impl TryFrom<&Certificate> for PckTcbInfo {
+    type Error = Error;
+
+    fn try_from(cert: &Certificate) -> Result<Self, Self::Error> {
+        let sgx_extensions = sgx_extensions(cert)?;
+
+        let fmspc = fmspc(&sgx_extensions)?;
+
+        let (pce_svn, svns) = tcb_svns(&sgx_extensions)?;
+
+        Ok(PckTcbInfo {
+            svns,
+            pce_svn,
+            fmspc,
+        })
+    }
+}
+
+/// Get the [`SgxExtensions`] from the `cert`.
+///
+/// # Errors
+/// * `Error::MissingSgxExtension` if the `cert` does not have the SGX extension.
+/// * `Error::DerDecoding` if the contained DER is invalid.
+fn sgx_extensions(cert: &Certificate) -> Result<SgxExtensions, Error> {
+    let extensions = &cert.tbs_certificate.extensions;
+    let extension = match extensions
+        .iter()
+        .flatten()
+        .find(|extension| extension.extn_id == SGX_OID)
+    {
+        Some(extension) => extension,
+        None => return Err(Error::MissingSgxExtension(SGX_OID.to_string())),
+    };
+
+    let der_bytes = extension.extn_value.as_bytes();
+    Ok(SgxExtensions::from_der(der_bytes)?)
+}
+
+/// Get the FMSPC value from the extensions
+///
+/// # Errors
+/// * `Error::MissingSgxExtension` if the `cert` does not have the FMSPC extension.
+/// * `Error::DerDecoding` if the FMSPC DER value is not an OctetString
+/// * `Error::FmspcSize` if the FMSPC DER value is not exactly 6 bytes.
+fn fmspc(sgx_extensions: &SgxExtensions) -> Result<[u8; FMSPC_SIZE], Error> {
+    let fmspc_value = oid_value(&FMSPC_OID, sgx_extensions)?;
+    let octet = fmspc_value.decode_as::<OctetStringRef>()?;
+    let fmspc_bytes = octet.as_bytes();
+
+    if fmspc_bytes.len() != FMSPC_SIZE {
+        return Err(Error::FmspcSize(fmspc_bytes.len()));
+    }
+
+    let mut fmspc = [0u8; FMSPC_SIZE];
+    fmspc.copy_from_slice(fmspc_bytes);
+    Ok(fmspc)
+}
+
+/// Get the value for the `oid`s attribute.
+///
+/// # Errors
+/// `Error::MissingSgxExtension` if the `oid` is not present in `extensions`.
+fn oid_value(oid: &ObjectIdentifier, extensions: &SgxExtensions) -> Result<AttributeValue, Error> {
+    match extensions.iter().find(|extension| &extension.oid == oid) {
+        Some(extension) => Ok(extension.value.clone()),
+        None => Err(Error::MissingSgxExtension(oid.to_string())),
+    }
```

Comment:
Same, I think you can use ok_or_else which is a bit less typing

Reply:
:+1: